### ISSUE TITLE
extra-bundled-libraries problem: cannot find existing library

### ISSUE THREAD

**Issue by bapcyk** [haskell/cabal#8826]
Guys, maybe it's a bug but maybe it's just a question.

I try to link with a dynamic library written in Go (I have not it's source and it's located in the project sub-directory as `clib/mylib/libmylib.so`). After a lot of investigation and experiments I found that there is an option `extra-bundled-libraries:` which could work, but it does not. It reports me an error that the cabal cannot find such library. I followed the naming convention but without any success: it reported a list of possible (but missing) names and one of them **actually existed on the disk**! After usage of `strace` I found that cabal tries to find it in the directory `$MY_PROJECT/dist-newstyle/build/x86_64-linux/ghc-9.2.3/project-0.1.0.0/build/` with a name `libmylib-ghc9.2.3.so` ! Why in this directory?

Well, I copied the `libmylib.so` to `$PROJECT_ROOT/dist-newstyle/build/x86_64-linux/ghc-9.2.3/project-0.1.0.0/build/libmylib-ghc9.2.3.so` and it started to fail differently: that the linked cannot find **mylib** : "-lmylib: No such file or directory" now.

So, I created a symbolic link in the same directory, so `libmylib.so -> libmylib-ghc9.2.3.so` and voila, now cabal is able to process the directive:

```
...
extra-bundled-libraries:
  mylib
...
```

and I get valid executable.

My environment:

```
OS: Linux OpenSUSE
LSB: LSB_VERSION="core-2.0-noarch:core-3.2-noarch:core-4.0-noarch:core-2.0-x86_64:core-3.2-x86_64:core-4.0-x86_64"
Cabal: 3.8.1.0
GHC: 9.2.3
```

I have to refer mylib in a platform independent way and independently on particular file system paths (it will be in the repository and other users use different locations in their systems).

So, is it a bug?

- Or I have to copy mylib dynamic library to this directory really under 2 names (with versioned GHC suffix and without it)? 
- Is there a more convenient way to do it?
- What to do in the case of Windows?
- If I will edit on the fly the cabal file (with Makefile or similar) than how to find the directory where the library should be located? Something with `cabal exec ghc-pkg` ? Or there is an another way?

**Comment by angerman** [comment]
This option is primarily for the case where you want to copy buildartifacts alongside the cabal products, Hoover cabal did not build the library but an external tool did.

I would believe it would also take the non -ghc$ver suffix'ed library, as a fallback. But yes, it expects that library in the build location and will copy it into tue cabal store as well (e.g. as part of the package).

This feature was initially added to support the need of GHC ljbraries that bundled a libCffi.a/.so/.dylib that wasn't built by cabal but some other build tool, however, the library should end up in the package in the cabal store for easy reference.

**Comment by topcoder1208** [comment]
```
source-repository-package
  type: git
  location: https://github.com/input-output-hk/io-sim
  tag: f4183f274d88d0ad15817c7052df3a6a8b40e6dc
  --sha256: 0vb2pd9hl89v2y5hrhrsm69yx0jf98vppjmfncj2fraxr3p3lldw
  subdir:
    io-classes
    io-sim
    strict-stm

source-repository-package
  type: git
  location: https://github.com/well-typed/plutonomy.git
  tag: 6c01302ba8cf3be4f71617e106cd5ef7ed10fc63
  --sha256: 
```
How can I get --sha256 for plutonomy?

**Comment by angerman** [comment]
> ```
> 
> source-repository-package
> 
>   type: git
> 
>   location: https://github.com/input-output-hk/io-sim
> 
>   tag: f4183f274d88d0ad15817c7052df3a6a8b40e6dc
> 
>   --sha256: 0vb2pd9hl89v2y5hrhrsm69yx0jf98vppjmfncj2fraxr3p3lldw
> 
>   subdir:
> 
>     io-classes
> 
>     io-sim
> 
>     strict-stm
> 
> 
> 
> source-repository-package
> 
>   type: git
> 
>   location: https://github.com/well-typed/plutonomy.git
> 
>   tag: 6c01302ba8cf3be4f71617e106cd5ef7ed10fc63
> 
>   --sha256: 
> 
> ```
> 
> How can I get --sha256 for plutonomy?

`nix-prefetch-git https://github.com/well-typed/plutonomy.git 6c01302ba8cf3be4f71617e106cd5ef7ed10fc63` should provide.

However this is a Haskell.nix topic and has nothing to do with cabal. `--sha256:` is technically a comment.

**Comment by topcoder1208** [comment]
yeah, thank you

**Comment by topcoder1208** [comment]
Where can I ask about this issue?
If anyone know an way to resolve this issue, please help me. when I run `nix-shell`, I have got this error. 
![image](https://user-images.githubusercontent.com/61309193/222885748-f9ce2dfb-9fa2-4980-9682-afaddceed258.png)

**Comment by angerman** [comment]
> Where can I ask about this issue?
> 
> If anyone know an way to resolve this issue, please help me. when I run `nix-shell`, I have got this error. 
> 
> ![image](https://user-images.githubusercontent.com/61309193/222885748-f9ce2dfb-9fa2-4980-9682-afaddceed258.png)
> 
> 

GitHub.com/input-output-hk/haskell.nix would be the appropriate repository.

**Comment by bapcyk** [comment]
> This option is primarily for the case where you want to copy buildartifacts alongside the cabal products, Hoover cabal did not build the library but an external tool did.
> 
> I would believe it would also take the non -ghc$ver suffix'ed library, as a fallback. But yes, it expects that library in the build location and will copy it into tue cabal store as well (e.g. as part of the package).
> 
> This feature was initially added to support the need of GHC ljbraries that bundled a libCffi.a/.so/.dylib that wasn't built by cabal but some other build tool, however, the library should end up in the package in the cabal store for easy reference.

yep. It would be good to have ability to link any .so/.dll from any (relative as well) folder. And to run `cabal test` with some option so cabal would be able to pass the library in the way similar to LD_LIBRARY_PATH but in cross-platform way, IMHO :) Maybe there is a way, cos Idk?
